reader.go

```diff
@@ -705,15 +710,116 @@ func (d *decoder) parseDataFields(dm *defmsg, knownMsg bool, msgv reflect.Value)
 		if err != nil {
 			return reflect.Value{}, fmt.Errorf("error parsing data developer message: %v (field %d [%v] for [%v])", err, i, ddfd, dm)
 		}
+		fieldDescMap, ok := d.fieldDescMsgs[ddfd.devDataIndex]
+		if !ok {
+			if d.debug {
+				d.opts.logger.Printf("could not find dev field description for dev data index %d dev field #%d", ddfd.devDataIndex, ddfd.fieldNum)
+			}
+			continue
+		}
+		fieldDesc, ok := fieldDescMap[ddfd.fieldNum]
+		if !ok {
+			if d.debug {
+				d.opts.logger.Printf("could not find dev field description for dev field #%d, dm=%v", ddfd.fieldNum, dm)
+			}
+			continue
+		}
+		value, err := d.parseDevField(dm, fieldDesc, ddfd)
+		if err != nil {
+			return reflect.Value{}, err
+		}
+		var fieldName string
+		if len(fieldDesc.FieldName) > 0 {
+			fieldName = fieldDesc.FieldName[0]
+		} else {
+			fieldName = ""
+		}
+		var units string
+		if len(fieldDesc.Units) > 0 {
+			units = fieldDesc.Units[0]
+		} else {
+			units = ""
+		}
+
+		devField := DeveloperField{
+			DeveloperDataIndex:    fieldDesc.DeveloperDataIndex,
+			FieldDefinitionNumber: fieldDesc.FieldDefinitionNumber,
+			BaseTypeId:            uint8(fieldDesc.FitBaseTypeId),
+			FieldName:             fieldName,
+			Units:                 units,
+			Value:                 value,
+		}
+		devFieldsMap := msgv.FieldByName("DeveloperFields")
+		devFieldsMap.SetMapIndex(reflect.ValueOf(fieldName), reflect.ValueOf(devField))
 	}
 
 	if knownMsg && !msgv.IsValid() {
 		panic("internal decoder error: parse data fields: known message, but not (reflect) valid")
 	}
 
+	if msgv.IsValid() && msgv.Type() == reflect.TypeOf(FieldDescriptionMsg{}) {
+		fieldMsg := msgv.Interface().(FieldDescriptionMsg)
+		if d.fieldDescMsgs == nil {
+			d.fieldDescMsgs = map[byte]map[byte]FieldDescriptionMsg{}
+		}
+		fieldDescMap := d.fieldDescMsgs[fieldMsg.DeveloperDataIndex]
+		if fieldDescMap == nil {
+			fieldDescMap = map[byte]FieldDescriptionMsg{}
+		}
+		fieldDescMap[fieldMsg.FieldDefinitionNumber] = fieldMsg
+		d.fieldDescMsgs[fieldMsg.DeveloperDataIndex] = fieldDescMap
+	}
+
 	return msgv, nil
 }
 
+// unfortunately largely duplicated with parseFitField(). however, in this case, we do not know the type to instantiate
+// into a reflect.Value until we switch on the base type (and the base type in the FieldDescriptionMsg is different),
+// so this is how it is.
+func (d *decoder) parseDevField(dm *defmsg, fieldDescMsg FieldDescriptionMsg, devFieldDesc devDataFieldDesc) (reflect.Value, error) {
```

Comment:
any way to move most of this to a common func between this and parseFitField by passing in defmsg, size?

Reply:
Not that I can tell which is what I was trying to explain in the comment. The types they are switching on are different, and `parseFitField()` already has a `reflect.Value` instantiated with the proper backing value because it's writing to a typed struct (one of the generated messages). In this case we don't know the type to instantiate until we do the switch on the base type, so we need to d `reflect.ValueOf()` instead of `reflect.SetUint()` or similar.

Reply:
In short the only thing in common is going from bytes -> uint32, uint64, etc., which didn't seem worth breaking out.